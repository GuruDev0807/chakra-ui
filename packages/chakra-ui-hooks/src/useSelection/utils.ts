interface GetNextIndex {
  step?: number;
  currentIndex: number;
  itemsLength: number;
  loop: boolean;
}

export function getNextIndex({
  step = 1,
  currentIndex,
  itemsLength,
  loop,
}: GetNextIndex) {
  if (currentIndex === -1) {
    return step > 0 ? 0 : itemsLength - 1;
  }

  const nextIndex = currentIndex + step;
  if (nextIndex < 0) {
    return loop ? itemsLength - 1 : 0;
  }

  if (nextIndex >= itemsLength) {
    return loop ? 0 : itemsLength - 1;
  }

  return nextIndex;
}

export function getItemIndex<T>(items: T[], item: T) {
  if (item == null || items.length === 0) {
    return -1;
  }
  return items.indexOf(item);
}

export function getValue<T>(valueProp: T, valueState: T) {
  const isControlled = valueProp != null;
  const value = isControlled ? valueProp : valueState;
  return [value, isControlled] as [T, boolean];
}
