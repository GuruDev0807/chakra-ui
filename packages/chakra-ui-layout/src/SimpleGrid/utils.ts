import { SystemProps } from "@chakra-ui/system";
import { ResponsiveValue } from "styled-system";

// These helper fns are modified versions of the amazing rebass library
// https://github.com/rebassjs/rebass/blob/master/packages/layout/src/index.js

const px = (n: string | number) => (typeof n === "number" ? n + "px" : n);

export function widthToColumns(
  width: SystemProps["width"],
): SystemProps["gridTemplateColumns"] | null {
  if (Array.isArray(width)) {
    return width.map(widthToColumns as any);
  }

  if (
    width !== null &&
    typeof width === "object" &&
    Object.keys(width).length > 0
  ) {
    const acc: Record<string, string> = {};
    for (let key in width) {
      acc[key] = `repeat(auto-fit, minmax(${px(width[key])}, 1fr))`;
    }
    return acc;
  }

  if (width != null) {
    return `repeat(auto-fit, minmax(${px(width as any)}, 1fr))`;
  }

  return null;
}

export function countToColumns(
  count: ResponsiveValue<number> | undefined,
): SystemProps["gridTemplateColumns"] | null {
  if (Array.isArray(count)) {
    return count.map(countToColumns as any);
  }

  if (
    count !== null &&
    typeof count === "object" &&
    Object.keys(count).length > 0
  ) {
    const acc: Record<string, string> = {};
    for (let key in count) {
      acc[key] = `repeat(${count[key]}, 1fr)`;
    }
    return acc;
  }

  if (count != null) {
    return `repeat(${count}, 1fr)`;
  }

  return null;
}
