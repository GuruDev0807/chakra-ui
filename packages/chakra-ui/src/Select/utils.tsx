export function registerOption(items: any[], item: any) {
  if (items.length === 0) {
    return [item];
  }
  const index = items.findIndex(_item => _item.id === item.id);
  if (index >= 0) return items;

  const indexAfter = items.findIndex(
    _item =>
      !!(
        _item.ref.current &&
        item.ref.current &&
        item.ref.current.compareDocumentPosition(item.ref.current) &
          Node.DOCUMENT_POSITION_PRECEDING
      ),
  );

  if (indexAfter === -1) {
    return [...items, item];
  }

  return [...items.slice(0, indexAfter), item, ...items.slice(indexAfter)];
}
