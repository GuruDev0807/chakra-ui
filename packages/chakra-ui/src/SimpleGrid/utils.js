// These helper fns are modified versions of the amazing rebass library
// https://github.com/rebassjs/rebass/blob/master/packages/layout/src/index.js

const px = n => (typeof n === "number" ? n + "px" : n);

export const widthToColumns = width => {
  if (Array.isArray(width)) {
    return width.map(widthToColumns);
  }

  if (
    width !== null &&
    typeof width === "object" &&
    Object.keys(width).length > 0
  ) {
    const acc = {};
    for (let key in width) {
      acc[key] = `repeat(auto-fit, minmax(${px(width[key])}, 1fr))`;
    }
    return acc;
  }

  if (width != null) {
    return `repeat(auto-fit, minmax(${px(width)}, 1fr))`;
  }

  return null;
};

export const countToColumns = count => {
  if (Array.isArray(count)) {
    return count.map(countToColumns);
  }

  if (
    count !== null &&
    typeof count === "object" &&
    Object.keys(count).length > 0
  ) {
    const acc = {};
    for (let key in count) {
      acc[key] = `repeat(${count[key]}, 1fr)`;
    }
    return acc;
  }

  if (count != null) {
    return `repeat(${count}, 1fr)`;
  }

  return null;
};
