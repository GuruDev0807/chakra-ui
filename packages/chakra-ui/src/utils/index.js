import { useMemo, useLayoutEffect, useEffect } from "react";

export const assignRef = (ref, value) => {
  if (ref == null) return;
  if (typeof ref === "function") {
    ref(value);
  } else {
    try {
      ref.current = value;
    } catch (error) {
      throw new Error(`Cannot assign value "${value}" to ref "${ref}"`);
    }
  }
};

const focusableElList = [
  "a[href]",
  "area[href]",
  "button:not([disabled])",
  "embed",
  "iframe",
  "input:not([disabled])",
  "object",
  "select:not([disabled])",
  "textarea:not([disabled])",
  "*[tabindex]:not([aria-disabled])",
  "*[contenteditable]",
];

const focusableElSelector = focusableElList.join();

export function getFocusables(element, keyboardOnly = false) {
  let focusableEls = Array.from(element.querySelectorAll(focusableElSelector));

  // filter out elements with display: none
  focusableEls = focusableEls.filter(
    focusableEl => window.getComputedStyle(focusableEl).display !== "none",
  );

  if (keyboardOnly === true) {
    focusableEls = focusableEls.filter(
      focusableEl => focusableEl.getAttribute("tabindex") !== "-1",
    );
  }

  return focusableEls;
}

/// Evaluate color in theme object

const colorKeyInTheme = (theme, color) => color in theme.colors;

const colorHueValue = (theme, color) => {
  let hasDot = color.search(".") !== -1;
  if (hasDot) {
    const [colorName, hue] = color.split(".");

    if (colorKeyInTheme(theme, colorName)) {
      return theme.colors[colorName][hue];
    }
  }
  return null;
};

export const getColorInTheme = (theme, color) => {
  if (colorKeyInTheme(theme, color)) {
    return theme.colors[color][500];
  }

  if (colorHueValue(theme, color)) {
    return colorHueValue(theme, color);
  }

  return color;
};

export function setRef(ref, value) {
  if (typeof ref === "function") {
    ref(value);
  } else if (ref) {
    ref.current = value;
  }
}

export function useForkRef(refA, refB) {
  return useMemo(() => {
    if (refA == null && refB == null) {
      return null;
    }
    return refValue => {
      setRef(refA, refValue);
      setRef(refB, refValue);
    };
  }, [refA, refB]);
}

export function createChainedFunction(...funcs) {
  return funcs.reduce(
    (acc, func) => {
      if (func == null) {
        return acc;
      }

      return function chainedFunction(...args) {
        acc.apply(this, args);
        func.apply(this, args);
      };
    },
    () => {},
  );
}

export const useEnhancedEffect =
  typeof window !== "undefined" ? useLayoutEffect : useEffect;

export const wrapEvent = (theirHandler, ourHandler) => event => {
  if (theirHandler) {
    theirHandler(event);
  }

  if (!event.defaultPrevented) {
    return ourHandler(event);
  }
};

export const isReducedMotion = () => {
  const { matches } = window.matchMedia("(prefers-reduced-motion: reduce)");
  return matches;
};

export const prefersReducedMotion = () => ({
  "@media (prefers-reduced-motion: reduce)": {
    animation: "none",
    transition: "none",
  },
});
