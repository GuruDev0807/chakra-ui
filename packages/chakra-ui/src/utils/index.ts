export * from "./utils";

export type Merge<T, V> = Omit<T, Extract<keyof T, keyof V>> & V;
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type Void = () => void;

export type Dictionary<T = any> = {
  [key: string]: T;
};

export type Required<T> = {
  [P in keyof T]-?: T[P];
};
