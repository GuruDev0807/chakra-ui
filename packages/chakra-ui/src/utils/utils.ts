import {
  useMemo,
  useLayoutEffect,
  useEffect,
  MutableRefObject,
  Ref,
} from "react";
import { canUseDOM } from "exenv";

export const assignRef = (ref: Ref<any>, value: any) => {
  if (ref == null) return;
  if (typeof ref === "function") {
    ref(value);
  } else {
    try {
      (ref as MutableRefObject<any>).current = value;
    } catch (error) {
      throw new Error(`Cannot assign value "${value}" to ref "${ref}"`);
    }
  }
};

const focusableElList = [
  "a[href]",
  "area[href]",
  "button:not([disabled])",
  "embed",
  "iframe",
  "input:not([disabled])",
  "object",
  "select:not([disabled])",
  "textarea:not([disabled])",
  "*[tabindex]:not([aria-disabled])",
  "*[contenteditable]",
];

const focusableElSelector = focusableElList.join();

export function getFocusables(domNode: HTMLElement, keyboardOnly = false) {
  let focusableEls: HTMLElement[] = Array.from(
    domNode.querySelectorAll(focusableElSelector),
  );

  // filter out elements with display: none
  focusableEls = focusableEls.filter(
    focusableEl => window.getComputedStyle(focusableEl).display !== "none",
  );

  if (keyboardOnly === true) {
    focusableEls = focusableEls.filter(
      focusableEl => focusableEl.getAttribute("tabindex") !== "-1",
    );
  }

  return focusableEls;
}

export function useRefs(refA: Ref<any>, refB: Ref<any>) {
  return useMemo(() => {
    if (refA == null && refB == null) {
      return null;
    }
    return (refValue: any) => {
      assignRef(refA, refValue);
      assignRef(refB, refValue);
    };
  }, [refA, refB]);
}

export function createChainedFunction(...funcs: any[]) {
  return funcs.reduce(
    (acc, func) => {
      if (func == null) {
        return acc;
      }

      return function chainedFunction(this: any, ...args: any[]) {
        acc.apply(this, args);
        func.apply(this, args);
      };
    },
    () => {},
  );
}

export const useEnhancedEffect = canUseDOM ? useEffect : useLayoutEffect;

export function wrapEvent<T extends (event: any) => void>(
  theirHandler: T,
  ourHandler: T,
) {
  return function(event: Parameters<T>[0]) {
    if (typeof theirHandler === "function") {
      theirHandler(event);
    }

    if (!event.defaultPrevented && typeof ourHandler === "function") {
      return ourHandler(event);
    }
  };
}

export const isReducedMotion = () => {
  const { matches } = window.matchMedia("(prefers-reduced-motion: reduce)");
  return matches;
};

export const prefersReducedMotion = () => ({
  "@media (prefers-reduced-motion: reduce)": {
    animation: "none",
    transition: "none",
  },
});

export const isObject = (input: any) =>
  input != null && typeof input === "object" && Object.keys(input).length > 0;

export const inputProps = [
  "name",
  "type",
  "autoFocus",
  "size",
  "form",
  "pattern",
  "placeholder",
  "onBlur",
  "onChange",
  "onKeyDown",
  "onKeyUp",
  "onKeyPress",
  "onFocus",
  "id",
  "autoFocus",
  "aria-label",
  "aria-describedby",
  "aria-labelledby",
  "min",
  "max",
  "maxLength",
  "minLength",
  "step",
  "defaultValue",
  "value",
  "isReadOnly",
  "isFullWidth",
  "isDisabled",
  "isInvalid",
  "isRequired",
];
