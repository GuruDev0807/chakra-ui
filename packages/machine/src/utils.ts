import { Dict, StateMachine as S } from "./types"

export const isString = (value: any): value is string =>
  typeof value === "string"

export const isArray = (value: any): value is any[] => Array.isArray(value)

export const isObject = (value: any): value is Dict =>
  value != null && typeof value === "object" && !isArray(value)

export const isFunction = (value: any): value is Function =>
  typeof value === "function"

export const keys = <T extends Record<string, any>>(value: T) =>
  Object.keys(value) as string[]

export const toArray = (v: any | any[] = []) => (isArray(v) ? v : [v])

export function executeActions<
  TContext extends Dict,
  TEvent extends S.EventObject
>(context: TContext, event: TEvent, actions?: S.Actions<TContext, TEvent>) {
  if (!actions) return
  actions = toArray(actions)
  for (const action of actions) {
    action(context, event)
  }
}

export function toEvent<T extends string | S.Event>(event: T) {
  return isString(event) ? ({ type: event } as S.EventObject) : event
}

export function toTarget<T extends string | { target?: string }>(target: T) {
  return isString(target) ? { target } : target
}

export function toTransition<
  TContext extends Dict,
  TState extends string,
  TEvent extends S.EventObject
>(
  transition: S.Transitions<TContext, TState, TEvent> | undefined,
  currentState?: string,
) {
  const _transition = isString(transition) ? toTarget(transition) : transition

  const check = (t: S.TransitionDefinition<TContext, string, TEvent>) => {
    const isTargetless = t.actions && !t.target
    if (isTargetless && currentState) t.target = currentState
    return t
  }

  if (isArray(_transition)) {
    return _transition.map(check)
  }

  if (isObject(_transition)) {
    return check(_transition)
  }
}

export function pickTarget<
  TContext extends Dict,
  TState extends string,
  TEvent extends S.EventObject
>(
  context: Dict,
  event: TEvent,
  transitions?: S.Transitions<TContext, TState, TEvent>,
): S.TransitionDefinitionWithDelay<TContext, TState, TEvent> | undefined {
  return toArray(transitions).find((t) => {
    const transition = toTarget(t)
    return transition.cond?.(context, event) ?? transition.target
  })
}

export function createEveryActivities<
  TContext extends Dict,
  TState extends string,
  TEvent extends S.EventObject
>(
  every: S.StateNode<TContext, TState, TEvent>["every"],
  iterator: (activity: S.Activity<TContext, TEvent>) => void,
) {
  if (!every) return
  for (const timeout in every) {
    const activity = (context: TContext, event: TEvent) => {
      const actions = every?.[timeout]
      const id = setInterval(() => {
        executeActions(context, event, actions as S.Actions<TContext, TEvent>)
      }, +timeout)
      return () => clearInterval(id)
    }
    iterator(activity)
  }
}

export function executeActivities<
  TContext extends Dict,
  TEvent extends S.EventObject
>(
  context: TContext,
  event: TEvent,
  activities: S.Activities<TContext, TEvent>,
  iterator: (fn: VoidFunction) => void,
) {
  if (isArray(activities)) {
    for (const activity of activities) {
      const fn = activity(context, event)
      iterator(fn)
    }
  } else {
    const fn = activities?.(context, event)
    iterator(fn)
  }
}
