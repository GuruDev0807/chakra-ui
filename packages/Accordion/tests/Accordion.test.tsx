import "@testing-library/jest-dom/extend-expect"
import userEvent from "@testing-library/user-event"
import React from "react"
import { render, fireEvent } from "../../../test/utils"
import {
  BaseAccordion,
  BaseAccordionButton,
  BaseAccordionItem,
  BaseAccordionPanel,
} from "../src/Accordion.base"

test("Button renders correctly", () => {
  const { asFragment } = render(
    <BaseAccordion>
      <BaseAccordionItem>
        <BaseAccordionButton>Section 1 title</BaseAccordionButton>
        <BaseAccordionPanel>Panel 1</BaseAccordionPanel>
      </BaseAccordionItem>
    </BaseAccordion>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("uncontrolled: It opens the accordion panel", () => {
  const { getByTestId } = render(
    <BaseAccordion defaultIndex={0}>
      <BaseAccordionItem>
        <BaseAccordionButton data-testid="button">
          Section 1 title
        </BaseAccordionButton>
        <BaseAccordionPanel data-testid="panel">Panel 1</BaseAccordionPanel>
      </BaseAccordionItem>
    </BaseAccordion>,
  )

  const button = getByTestId("button")
  expect(button).toHaveAttribute("aria-expanded", "true")
})

test("uncontrolled: toggles the accordion on click", () => {
  const { getByText } = render(
    <BaseAccordion>
      <BaseAccordionItem>
        <BaseAccordionButton>Trigger</BaseAccordionButton>
        <BaseAccordionPanel>Panel</BaseAccordionPanel>
      </BaseAccordionItem>
    </BaseAccordion>,
  )

  const trigger = getByText("Trigger")

  userEvent.click(trigger)
  expect(trigger).toHaveAttribute("aria-expanded", "true")

  // you can't toggle an accordion without passing `allowToggle`
  userEvent.click(trigger)
  expect(trigger).toHaveAttribute("aria-expanded", "true")
})

// test that arrow up & down moves focus to next/previous accordion
test("arrow up & down moves focus to next/previous accordion", () => {
  const { getByText, getByTestId } = render(
    <BaseAccordion data-testid="accordion">
      <BaseAccordionItem>
        <BaseAccordionButton>Section 1 title</BaseAccordionButton>
        <BaseAccordionPanel>Panel 1</BaseAccordionPanel>
      </BaseAccordionItem>

      <BaseAccordionItem>
        <BaseAccordionButton>Section 2 title</BaseAccordionButton>
        <BaseAccordionPanel>Panel 2</BaseAccordionPanel>
      </BaseAccordionItem>
    </BaseAccordion>,
  )

  const accordion = getByTestId("accordion")
  const accordionItemOne = getByText("Section 1 title")
  const accordionItemTwo = getByText("Section 2 title")

  fireEvent.keyDown(accordion, { key: "ArrowDown", keyCode: 40 })
  expect(accordionItemOne).toHaveFocus()

  fireEvent.keyDown(accordion, { key: "ArrowUp", keyCode: 38 })
  expect(accordionItemTwo).toHaveFocus()
})

// test that home & end keys moves focus to first/last accordion
test("home & end keys moves focus to first/last accordion", () => {
  const { getByText, getByTestId } = render(
    <BaseAccordion data-testid="accordion">
      <BaseAccordionItem>
        <BaseAccordionButton>First section</BaseAccordionButton>
        <BaseAccordionPanel>Panel 1</BaseAccordionPanel>
      </BaseAccordionItem>

      <BaseAccordionItem>
        <BaseAccordionButton>Second section</BaseAccordionButton>
        <BaseAccordionPanel>Panel 1</BaseAccordionPanel>
      </BaseAccordionItem>

      <BaseAccordionItem>
        <BaseAccordionButton>Last section</BaseAccordionButton>
        <BaseAccordionPanel>Panel 2</BaseAccordionPanel>
      </BaseAccordionItem>
    </BaseAccordion>,
  )
  const accordion = getByTestId("accordion")
  const first = getByText("First section")
  const last = getByText("Last section")

  fireEvent.keyDown(accordion, { key: "Home", keyCode: 36 })
  expect(first).toHaveFocus()

  fireEvent.keyDown(accordion, { key: "End", keyCode: 35 })
  expect(last).toHaveFocus()
})
// test the only one accordion can be visible + is not togglable
// test the only one accordion can be visible + is togglable
// test that multiple accordions can be opened + is togglable
// it has the proper aria attributes
// test that enter and space can toggle the visiblity
// test that tab moves focus to the next focusable element;
// test that aria-contols for button is same as id for panel
// test that aria-expanded is true/false when accordion is open/closed
// test that panel has role=region and aria-labelledby
