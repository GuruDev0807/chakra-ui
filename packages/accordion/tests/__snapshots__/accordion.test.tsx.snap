// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-accordion css-0"
  >
    <div
      class="chakra-accordion__item css-1wbdjbz"
    >
      <button
        aria-controls="accordion-panel-1"
        aria-expanded="false"
        class="chakra-accordion__button css-q4rc4p"
        id="accordion-button-1"
      >
        Section 1 title
      </button>
      <div
        hidden=""
        id="collapse-mock"
      >
        <div
          aria-labelledby="accordion-button-1"
          class="chakra-accordion__panel css-wxchgh"
          id="accordion-panel-1"
          role="region"
        >
          Panel 1
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
